zproject/backends.py

```diff
@@ -352,6 +352,7 @@ def authenticate(self, request: HttpRequest, *,
         if user_profile is None:
             return None
         if user_profile.check_password(password):
+            request.session['needs_to_change_password'] = not check_password_strength(password)
```

Comment:
Hmm, since `session` doesn't make much sense in some contexts, such as clients that operate purely on API requests and their session is useless (perhaps we might actually have requests without a session actually occutrrings at some point in the future?), I wonder if it wouldn't be better to do a conditional `if hasattr(request, 'session'`) here. And if there's no session and password is weak, raise an appropriate json error. Perhaps @timabbott will have some thoughts on this, but this would be my approach.  With this, we wouldn't need the weird `request.sessoin = self.client.session` things in tests.

Reply:
There's also one more weird scenario that we need to handle here. I though I talked about it in the topic in #backend, but I checked and apparently no :disappointed: So, if you dive into the `check_password` codepath in Django, there's a somewhat obscure situation that can happen behind the scenes - iirc, it'll be triggered if the password hashing algorithm is changed - in that case, when the user is logging in, Django first checks the password, if it's correct, Django checks the hash algorithm of the password hash in the database, and if it's the old algorithm, django will run `set_password` for this user, to update the hash to the new algorithm. Something like that. Here's a traceback this can generate if the password is also weak:
```
Traceback (most recent call last):
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "./zerver/views/auth.py", line 777, in login_page
    extra_context=extra_context, **kwargs)(request)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/decorators/debug.py", line 76, in sensitive_post_parameters_wrapper
    return view(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 142, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/views.py", line 61, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/generic/edit.py", line 141, in post
    if form.is_valid():
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 185, in is_valid
    return self.is_bound and not self.errors
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 180, in errors
    self.full_clean()
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 382, in full_clean
    self._clean_form()
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 409, in _clean_form
    cleaned_data = self.clean()
  File "./zerver/forms.py", line 341, in clean
    realm=realm, return_data=return_data)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/__init__.py", line 73, in authenticate
    user = backend.authenticate(request, **credentials)
  File "</srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/decorator.py:decorator-gen-2>", line 2, in authenticate
  File "./zproject/backends.py", line 222, in rate_limit_auth
    result = auth_func(*args, **kwargs)
  File "./zproject/backends.py", line 308, in authenticate
    if user_profile.check_password(password):
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/base_user.py", line 111, in check_password
    return check_password(raw_password, self.password, setter)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/hashers.py", line 61, in check_password
    setter(password)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/base_user.py", line 107, in setter
    self.set_password(raw_password)
  File "./zerver/models.py", line 1210, in set_password
    raise PasswordTooWeakError
zerver.models.PasswordTooWeakError
```

Long story short, `user_profile.check_password(password)` can raise `PasswordTooWeakError`. We need to handle that gracefully here somehow - since this is such a weird and rare scenario, perhaps an appropriate json error would be good enough? Not sure.

Reply:
Added a new commit for this case.